20221012-add-type-restrictions-to-dsl-syntax.md

```diff
@@ -0,0 +1,171 @@
+
+```
+model
+  schema 1.1
+```
+
+## Type restrictions
+
+### Type restrictions
+
+Let A be an authorization model with set of types T and set of relations R. A type restriction on A can be seen as a set-valued function from TxR to TxR*, where R* is R with the addition of an "empty" relation. That is, a type restriction takes any (type, relation) in A and takes it to the set of type-relation pairs (where the relation may be empty) that it may be directly related to. That is, if the pair (folder, viewer) has the type restriction (employee, \_), where "\_" is used to indicate the empty relation, then the only tuples of the form (object=folder:documents, relation=viewer, user=U) that are allowed are ones in which `U=employee:X`, where `X` is an employee id.
+
+#### Examples
+
+1. Suppose A is an authorization model with a type folder and relation parent on folder. We may have a type restriction on A that takes (folder, parent) to the set {(folder, \_)}, meaning that parents of folders must be folders.
+ 2. Suppose A is an authorization model with type document and relation reader on document. We may have a type restriction on A that takes (document, reader) to the set {(user, \_), (group, member)}, meaning that the only users that may be readers of a document are users with type user, or a userset of group members.
+
+### Syntax
+
+Let A be an authorization model with set of types T and set of relations R, and recall that R* is the set R with the addition of an "empty" relation. We will write an element (t, r) of TxR* as t#r. The special case (t, _) is simply written as t. If the type restriction takes (t, r) to a set {(t1, r1), (t2, r2)}, then in our model we will write this as
+```
+type t
+  relations
+    define r: [t1#r1, t2#r2] ...
+```
+If the type restriction takes (t, r) to the empty set we will simply write nothing in our model.
+
+### Types with no relations
+
+It now becomes useful to allow types with no relations. For example, suppose we want to define a `viewer` relation and retrict the typeswe may want to specify a `user` type which is the type for all concrete users. This will look like
+```
+type user
+```
+
+### The dropping of "as" and "self"
+
+In the new DSL we will drop "as" and "self". We drop "as" by adding ":" to all relations, even ones with no type restrictions. That is, if previously we had `define editor as owner`, this becomes `define editor: owner`. With the addition of type restrictions means we may also drop `self` from the DSL. This is because if a pair (type, relation) has a non-empty set of type restrictions, then the assumption is that (type, relation) is a pair that allows a direct relation to type. Thus, previously if we defined a relation as
```

Comment:
Mind if we say "we will drop the "as" and "self" keywords" (add keywords to the end)?

Reply:
Thanks!